### PR TITLE
storage_service api: handle dropped tables

### LINKED ISSUE #12007: Cleanup may fail if a table is dropped concurrently
Currently nodetool cleanup may fail if a table is dropped in the middle of cleanup.

For example, the failure looks like:
```
java.lang.IllegalStateException: Scylla API server HTTP POST to URL '/storage_service/keyspace_cleanup/ks' failed: std::_Nested_exception<no_such_column_family> $
Can't find a column family with UUID 01234567-89ab-cdef-0123-000000000000): std::out_of_range (_Map_base::at)
        at com.scylladb.jmx.api.APIClient.getException(APIClient.java:140)
        at com.scylladb.jmx.api.APIClient.post(APIClient.java:120)
        at com.scylladb.jmx.api.APIClient.post(APIClient.java:130)
        at com.scylladb.jmx.api.APIClient.post(APIClient.java:113)
        at com.scylladb.jmx.api.APIClient.postGetVal(APIClient.java:144)
        at com.scylladb.jmx.api.APIClient.postInt(APIClient.java:148)
        at org.apache.cassandra.service.StorageService.forceKeyspaceCleanup(StorageService.java:629)
        at org.apache.cassandra.service.StorageService.forceKeyspaceCleanup(StorageService.java:1665)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:72)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:276)
        at com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:112)
        at com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:46)
        at com.sun.jmx.mbeanserver.MBeanIntrospector.invokeM(MBeanIntrospector.java:237)
        at com.sun.jmx.mbeanserver.PerInterface.invoke(PerInterface.java:138)
        at com.sun.jmx.mbeanserver.MBeanSupport.invoke(MBeanSupport.java:252)
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
        at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
        at com.scylladb.jmx.utils.APIMBeanServer.invoke(APIMBeanServer.java:188)
        at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1468)
        at javax.management.remote.rmi.RMIConnectionImpl.access$300(RMIConnectionImpl.java:76)
        at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1309)
        at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1401)
        at javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:829)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:357)
        at sun.rmi.transport.Transport$1.run(Transport.java:200)
        at sun.rmi.transport.Transport$1.run(Transport.java:197)
        at java.security.AccessController.doPrivileged(Native Method)
        at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
        at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
        at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
        at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
        at java.security.AccessController.doPrivileged(Native Method)
        at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

While on the node log, we see the table being dropped:
```
Nov 17 10:11:12 ip-127-0-0-1 scylla[12345]:  [shard 1] schema_tables - Dropping ks.t_aabbccddef_2022_11_17_01_02_03 id=01234567-89ab-cdef-0123-000000000000 version=c3ace1cd-d84b-458e-8864-c88cbafaae77
```

### PR DESCRIPTION
**Pull request by bhalevy**
Gracefully skip tables that were removed in the background.

Fixes #12007

### CODE REVIEW

**Comment by bhalevy** [comment]
@Deexie this series touches areas that you last modified, please review too.

File: api/storage_service.cc [hunk 1]
```diff
@@ -609,92 +661,103 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
...
+            co_await ctx.db.invoke_on_all([&] (replica::database& db) -> future<> {
+                auto local_tables = table_infos;
                 // cleanup smaller tables first, to increase chances of success if low on space.
-                std::ranges::sort(table_ids, std::less<>(), [&] (const table_id& id) {
-                    return db.find_column_family(id).get_stats().live_disk_space_used;
+                std::ranges::sort(local_tables, std::less<>(), [&] (const table_info& ti) {
+                  try {
+                    return db.find_column_family(ti.id).get_stats().live_disk_space_used;
+                  } catch (const replica::no_such_column_family& e) {
+                    return int64_t(-1);
```
Comment (raphaelsc) [inline api/storage_service.cc:708]: bad indentation

File: api/storage_service.cc [hunk 2]
```diff
@@ -609,92 +661,103 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
...
 
-    ss::perform_keyspace_offstrategy_compaction.set(r, wrap_ks_cf(ctx, [] (http_context& ctx, std::unique_ptr<request> req, sstring keyspace, std::vector<sstring> tables) -> future<json::json_return_type> {
-        apilog.info("perform_keyspace_offstrategy_compaction: keyspace={} tables={}", keyspace, tables);
-        co_return co_await ctx.db.map_reduce0([&keyspace, &tables] (replica::database& db) -> future<bool> {
+    ss::perform_keyspace_offstrategy_compaction.set(r, wrap_ks_cf(ctx, [] (http_context& ctx, std::unique_ptr<request> req, sstring keyspace, std::vector<table_info> table_infos) -> future<json::json_return_type> {
+        apilog.info("perform_keyspace_offstrategy_compaction: keyspace={} tables={}", keyspace, table_infos);
+        bool res = false;
+        try {
+          res = co_await ctx.db.map_reduce0([&] (replica::database& db) -> future<bool> {
             bool needed = false;
```
Comment (raphaelsc) [inline api/storage_service.cc:730]: ditto.

File: api/storage_service.cc [hunk 3]
```diff
@@ -628,26 +630,25 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
...
-                        std::runtime_error("Can not perform cleanup operation when topology changes"));
-            }
-            return ctx.db.invoke_on_all([keyspace, column_families] (replica::database& db) -> future<> {
+        if (!co_await ss.local().is_cleanup_allowed(keyspace)) {
+            auto msg = "Can not perform cleanup operation when topology changes";
+            apilog.warn("force_keyspace_cleanup: keyspace={} tables={}: {}", keyspace, column_families, msg);
+            co_await coroutine::return_exception(std::runtime_error(msg));
+        }
+        // FIXME: log error
+            co_await ctx.db.invoke_on_all([&] (replica::database& db) -> future<> {
```
Comment (Deexie) [inline api/storage_service.cc:701]: Save the captures. `ctx` cannot be accessed here due to `co_await` in 645.

File: api/storage_service.cc [hunk 4]
```diff
@@ -681,20 +679,21 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
...
-            return do_for_each(column_families, [=, &db](sstring cfname) {
+            for (const auto& cfname : column_families) {
                 auto& cm = db.get_compaction_manager();
                 auto& cf = db.find_column_family(keyspace, cfname);
-                return cm.perform_sstable_upgrade(owned_ranges_ptr, cf.as_table_state(), exclude_current_version);
-            });
-        }).then([]{
-            return make_ready_future<json::json_return_type>(0);
-        });
+                co_await cm.perform_sstable_upgrade(owned_ranges_ptr, cf.as_table_state(), exclude_current_version);
```
Comment (Deexie) [inline api/storage_service.cc]: ditto

File: api/storage_service.cc [hunk 5]
```diff
@@ -609,14 +609,16 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
...
         auto column_families = parse_tables(keyspace, ctx, req->query_parameters, "cf");
         if (column_families.empty()) {
             column_families = map_keys(ctx.db.local().find_keyspace(keyspace).metadata().get()->cf_meta_data());
         }
         apilog.debug("force_keyspace_compaction: keyspace={} tables={}", keyspace, column_families);
-        return ctx.db.invoke_on_all([keyspace, column_families] (replica::database& db) -> future<> {
+        // FIXME: log error
+            // FIXME: indentation broken on purpose
+            co_await ctx.db.invoke_on_all([&] (replica::database& db) -> future<> {
             auto table_ids = boost::copy_range<std::vector<table_id>>(column_families | boost::adaptors::transformed([&] (auto& cf_name) {
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply api/storage_service.cc]: 2021.1 already supports coroutines
I don't think it's worth backporting to 2020

File: api/storage_service.cc (hunk 1 above)
Comment (bhalevy) [reply api/storage_service.cc:708]: yea, I tried to minimize churn

**Comment by bhalevy** [comment]
In v2 ([f79bac6](https://github.com/scylladb/scylladb/commit/f79bac6cae411ca71e3a2a7e6a614e58bc278cf8)):
- kept `auto& db = ctx.db` on the stack in coroutines before preemption
- fixed indentation

File: api/storage_service.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (denesb) [reply api/storage_service.cc]: 2021.1 has coroutines on experimental still, only enabled for raft.

File: api/storage_service.cc [hunk 6]
```diff
@@ -100,6 +108,34 @@ std::vector<sstring> parse_tables(const sstring& ks_name, http_context& ctx, con
     return parse_tables(ks_name, ctx, it->second);
 }
 
+std::vector<table_info> parse_table_infos(const sstring& ks_name, http_context& ctx, sstring value) {
+    std::vector<table_info> res;
+    if (value.empty()) {
+        const auto& cf_meta_data = ctx.db.local().find_keyspace(ks_name).metadata().get()->cf_meta_data();
```
Comment (denesb) [inline api/storage_service.cc:114]: Should probably catch `no_such_keyspace` here too.

File: api/storage_service.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bhalevy) [reply api/storage_service.cc]: I think it's low tisk to use them at the api layer as well.  We do nothing fancy their.

File: api/storage_service.cc (hunk 3 above)
Comment (bhalevy) [reply api/storage_service.cc]: Done in v2

File: api/storage_service.cc (hunk 6 above)
Comment (bhalevy) [reply api/storage_service.cc:114]: good idea

**Comment by bhalevy** [comment]
In v3 ([ec5707a](https://github.com/scylladb/scylladb/commit/ec5707a4a8040090ea8e62ff835d7a25831e475e)):
- rebased
- handle `no_such_keyspace` in `parse_table_infos`
  - (not needed in `run_on_existing_tables` since it calls find_uuid that may only throw `no_such_column_family`)

**Review by denesb — APPROVED** [review]

**Pull request closed without merging**